<h1><u>Data Wrangling - Coffee Shop</u>

<u>**Problem Statement:**</u>
<br>What three features most affect the price per ounce of coffee?

<u>**Context:**</u>
<br>A start-up coffee company is creating their signature blend to sell alongside the more generic blends of coffee. The start-up needs to know what three features to primarily incorporate into their signature blend to maximize its popularity and distinguish their company from other coffee companies.

<u>**Criteria for Success:**</u>
<br>Determine the three coffee features that will create a popular, signature blend of coffee.

<u>**Scope of Solution Space:**</u>
<br>Rating
<br>Acidity
<br>Aftertaste
<br>Aroma
<br>Body
<br>Flavor
<br>Review description
<br>Country of origin
<br>Roast level
<br>Roaster
<br>Roaster locatiion

<h3> Data Collection

My first step was to import os to access my local device operating system, and import pandas, matplotlib.pyplot, and seaborn to manage the data set and begin exploring the data.

Next I imported the dataset I found on Kaggle: https://www.kaggle.com/datasets/patkle/coffeereviewcom-over-7000-ratings-and-reviews
I initially recieved an error message, but found I had to use double slashes insetaed of single slashes in the CSV file address pathway for the data to import into my JupyterLab notebook.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [2]:
df = pd.read_csv("C:\\Users\\linds\\OneDrive\\Desktop\\LC_Springboard\\reviews_feb_2023.csv")

<h3> Data Definition

My first step in assessing the data cleaning needs was to view the first five entries of the data set using the head() fucntion.

In [3]:
print(df.head())

                                         title rating  acidity_structure  \
0                      Bolivia Manantial Gesha     93                9.0   
1  Yellow Pacamara Carbonic Maceration Nanolot     92                8.0   
2                   Ethiopia Gera Genji Challa     94                9.0   
3                 Yirgacheffe Mengesha Natural     94                9.0   
4           Tropical Summer Colombia La Sierra     93                9.0   

   aftertaste aroma body flavor with_milk agtron  \
0         8.0     9    8      9       NaN  60/78   
1         8.0     9    8      9       NaN  60/78   
2         8.0     9    9      9       NaN  59/77   
3         8.0     9    9      9       NaN  60/77   
4         8.0     9    8      9       NaN  60/77   

                                    blind_assessment  \
0  Richly aromatic, floral-toned. Magnolia, cocoa...   
1  Crisply sweet-savory. White peach, hop flowers...   
2  Delicately aromatic, complex. Lilac, cocoa nib...   
3  Hig

Already, it looks as though the with_milk column will be problematic due to null values. 
<br> This lead me to analyze the column .info() to see what type of information each category containted, and how many values were non-null.

In [4]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7041 entries, 0 to 7040
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              7041 non-null   object 
 1   rating             7041 non-null   object 
 2   acidity_structure  2166 non-null   float64
 3   aftertaste         6169 non-null   float64
 4   aroma              6991 non-null   object 
 5   body               7030 non-null   object 
 6   flavor             7025 non-null   object 
 7   with_milk          997 non-null    object 
 8   agtron             7041 non-null   object 
 9   blind_assessment   7040 non-null   object 
 10  bottom_line        2961 non-null   object 
 11  coffee_origin      6536 non-null   object 
 12  est_price          5002 non-null   object 
 13  notes              7033 non-null   object 
 14  review_date        7041 non-null   object 
 15  roast_level        6667 non-null   object 
 16  roaster            7041 

From the initilal info() reading, it is confirmed that the with_milk column has very few useable entries. I calculated the number of missing values in the each columm to further determine how much useable data is present. I found both the count of missing values and percentage of the column's data that represented missing values.

In [5]:
missing = pd.concat([df.isnull().sum(), 100 * df.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by=['count', '%'], ascending = False)

count          %
with_milk           6044  85.840080
acidity_structure   4875  69.237324
bottom_line         4080  57.946314
est_price           2039  28.958955
aftertaste           872  12.384604
coffee_origin        505   7.172277
roast_level          374   5.311745
aroma                 50   0.710126
flavor                16   0.227240
body                  11   0.156228
notes                  8   0.113620
roaster_location       3   0.042608
blind_assessment       1   0.014203
title                  0   0.000000
rating                 0   0.000000
agtron                 0   0.000000
review_date            0   0.000000
roaster                0   0.000000
url                    0   0.000000

I will drop the with_milk column since it is nearly 86% null values, and is not included in my problem statement's scope of solution. However, the acidity structure may prove important, and I need to decide what do to with the near 69% null values. 

In [6]:
df1 = df.drop('with_milk', axis=1, inplace = False)

In [7]:
print(df1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7041 entries, 0 to 7040
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              7041 non-null   object 
 1   rating             7041 non-null   object 
 2   acidity_structure  2166 non-null   float64
 3   aftertaste         6169 non-null   float64
 4   aroma              6991 non-null   object 
 5   body               7030 non-null   object 
 6   flavor             7025 non-null   object 
 7   agtron             7041 non-null   object 
 8   blind_assessment   7040 non-null   object 
 9   bottom_line        2961 non-null   object 
 10  coffee_origin      6536 non-null   object 
 11  est_price          5002 non-null   object 
 12  notes              7033 non-null   object 
 13  review_date        7041 non-null   object 
 14  roast_level        6667 non-null   object 
 15  roaster            7041 non-null   object 
 16  roaster_location   7038 

According to the info() print out, rating, acidity structure, aftertaste, aroma, body, and flavor columns should be numeric rather than object categories.
<br>I analyzed the unique values of each column to determine how many (if any) missing values the columns had, and and converted the column types to numeric forms so I could analyze the average of each feature.

In [8]:
df1['rating'].value_counts()

93    1368
92    1104
94     981
90     633
91     631
95     473
88     316
89     309
87     248
85     155
86     150
96     137
84      96
83      55
80      53
81      49
79      46
82      45
78      44
97      34
77      22
76      15
73      14
75      11
74       8
72       8
71       8
68       6
70       4
NR       4
67       3
98       3
63       2
65       2
69       1
66       1
52       1
60       1
Name: rating, dtype: int64

Since the rating category only had four entries that were not rated, I decided to drop these columns.

In [9]:
df1 = df1.drop(df1[df1['rating']=='NR'].index)

In [10]:
df1['rating'] = pd.to_numeric(df1['rating'], errors='coerce')
print(df1['rating'].dtype)

int64


In [11]:
df1['aroma'].value_counts()

9      3666
8      2292
7       619
6       164
10      111
5        55
4        21
3         6
5.8       6
6.3       4
5.1       4
6.1       4
5.6       4
5.7       3
5.4       3
7.2       3
2         3
4.8       2
5.2       2
7.6       2
7.5       2
6.2       2
4.3       2
7.4       1
4.7       1
6.5       1
6.4       1
5.9       1
4.5       1
4.9       1
Name: aroma, dtype: int64

In [12]:
df1['aroma'] = pd.to_numeric(df1['aroma'], errors='coerce')
print(df1['aroma'].dtype)

float64


In [13]:
df1['body'].value_counts()

8      3309
9      2401
7       966
6       230
5        59
4         7
10        5
5.7       5
5.3       5
5.8       4
6.4       4
5.9       3
5.5       3
5.2       3
5.1       3
5.6       3
6.3       2
7.1       2
5.4       2
6.1       2
6.2       2
6.7       1
7.3       1
6.6       1
4.8       1
4.3       1
4.9       1
Name: body, dtype: int64

In [14]:
df1['body'] = pd.to_numeric(df1['body'], errors='coerce')
print(df1['body'].dtype)

float64


In [15]:
df1['flavor'].value_counts()

9      4144
8      1711
7       565
10      243
6       170
5        89
4        29
3        13
5.3       6
5.7       5
5.8       4
7.4       3
6.5       3
6.1       3
6.7       3
6.9       3
2         3
7.3       2
5.6       2
3.7       2
7.7       2
8.1       1
5.5       1
4.6       1
4.8       1
5.1       1
4.9       1
4.4       1
1         1
7.1       1
3.9       1
6.2       1
6.4       1
4.2       1
5.2       1
6.8       1
5.4       1
Name: flavor, dtype: int64

In [16]:
df1['flavor'] = pd.to_numeric(df1['flavor'], errors='coerce')
print(df1['flavor'].dtype)

float64


ext, I analyzed the agtron, and est_price columns since I felt these should also be numeric.
<br>According to my original info() print out, agtron had ratio scoring, and the est_price column included price per unit verbage. Therefore, any missing values should be analyzed, but the category should not be changed.

In [17]:
df1['agtron'].value_counts()

/        257
58/76    146
58/78    125
60/78    124
0/0      107
        ... 
29/52      1
40/62      1
40/65      1
74/99      1
30/39      1
Name: agtron, Length: 1121, dtype: int64

The value_counts() function showed 257 entries with unuseable data (simply a / with no numbers). Since the second number in the ratio varied greatly, I did a google search of what an agtron rating was. It seems to be a ratio rating a coffee before and after grinding, so each unique ratio is a unique rating. I chose to drop the 257 unuseable entries, and leave the remaining data as-is for now.

Finally, I calculated the summary statistics for each of the numeric columns to give me a starting point as I explored the data to find which three features most affected the coffee blend's rating.

In [18]:
df1.describe().T

count       mean       std   min   25%   50%   75%   max
rating             7037.0  90.868126  4.121444  52.0  90.0  92.0  93.0  98.0
acidity_structure  2166.0   8.487535  0.629954   1.0   8.0   9.0   9.0  10.0
aftertaste         6169.0   7.968552  0.722743   2.0   8.0   8.0   8.0  10.0
aroma              6987.0   8.363189  0.912374   2.0   8.0   9.0   9.0  10.0
body               7026.0   8.095431  0.856112   4.0   8.0   8.0   9.0  10.0
flavor             7021.0   8.447728  0.997152   1.0   8.0   9.0   9.0  10.0